modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/operation/ComponentMessageProcessor.java

```diff
@@ -212,73 +246,89 @@ public CoreEvent process(CoreEvent event) throws MuleException {
 
   @Override
   public Publisher<CoreEvent> apply(Publisher<CoreEvent> publisher) {
-    Flux<CoreEvent> flux = from(publisher)
-        .flatMap(event -> subscriberContext().map(ctx -> addContextToEvent(event, ctx)));
-
-    if (isAsync()) {
-      final BiFunction<Throwable, Object, Throwable> localOperatorErrorHook =
-          getLocalOperatorErrorHook(this, errorTypeLocator, exceptionContextProviders);
-      return flux
-          // This flatMap allows the operation to run in parallel. The processing strategy relies on this
-          // (ProactorStreamProcessingStrategy#proactor) to do some performance optimizations.
-          .flatMap(event -> {
-            // Force the error mapper from the chain to be used.
-            // When using Mono.create with sink.error, the error mapper from the context is ignored, so it has to be
-            // explicitly used here.
-            DeferredMonoSinkExecutorCallback callback =
-                new DeferredMonoSinkExecutorCallback<>(t -> localOperatorErrorHook.apply(t, event));
-
-            onEvent(event, callback);
-
-            return create(callback::setSink);
-          });
-    } else {
-      return flux.handle((event, sink) -> {
-        try {
-          onEvent(event, new SynchronousSinkExecutorCallback(sink));
-        } catch (Throwable t) {
-          sink.error(unwrap(t));
-        }
-      });
-    }
+    final BiFunction<Throwable, Object, Throwable> localOperatorErrorHook =
+        getLocalOperatorErrorHook(this, errorTypeLocator, exceptionContextProviders);
+
+    return subscriberContext()
+        .flatMapMany(ctx -> {
+          final FluxSinkRecorder<Either<Throwable, CoreEvent>> errorSwitchSinkSinkRef = new FluxSinkRecorder<>();
+
+          Flux<CoreEvent> transformed = from(propagateCompletion(from(publisher), create(errorSwitchSinkSinkRef)
+              .map(result -> {
+                return result.reduce(me -> {
```

Comment:
add comments clarifying or extract to method with meaningful name, or both, or make a netflix documentary explaining this